### DISCUSSION TITLE
enhancement: console command for GPIO overview

### DISCUSSION THREAD

**Discussion by 0xdefec71f** [arendst/Tasmota#19810] [Ideas]
Hello,

I want to suggest an enhancement: It would be nice to have a console command for debugging purposes, that shows all (configured) inputs and outputs. So for example:

`getGPIO`

Output: `GPIO1=0, GPIO2=1, GPIO3=0, ...`

I'm currently struggling with a counter and noticed much too late, that the connected reed contact was permanently low.

Thank you and best regards!

**Comment by Wintespe** [comment]
Try **gpio all**

**Reply by sfromis** [reply]
Looks like the wish is for gpio pin input *values*, not just what is configured.

**Reply by 0xdefec71f** [reply]
right, it's about the values. I can see the configuration in the settings. For debug you need the values, regardless the configuration.

**Reply by 0xdefec71f** [reply]
Or maybe one could tweak the `gpio all` command and also show the values here beside the config.

**Reply by barbudor** [reply]
It doesn't necessarily makes sense.
How would it help to see the state of a GPIO that is defined as a DS18X20 sensor? Or a I2C pin or a software sensor?
If you temporarily want to see the level of a GPIO, set it as a Switch input

**Reply by 0xdefec71f** [reply]
For booleans it does make sense. No, a temporary config change does not help at all, because as mentioned above it would delete a counter value. For special functions one could skip it or show the analog value or a static label or whatever.

**Reply by s-hadinger** [reply]
I do like this feature which can be handy for debugging, but indeed there are some GPIOs that don't make sense to report

**Reply by 0xdefec71f** [reply]
I can't judge how complicated the implementation would be. But a raw digital value before being processed from the perspective of the firmware helps hunting down a layer 8 bug, without the need of reconfig or a multimeter. I also had the case of a ESP with defective input. The output could be raw-boolsch, raw-analog or raw-digital (e.g. DS18X20).

example: `GPIO1=true, GPIO2=255, GPIO3=0xFF88, ...`

Maybe I am the only one with this need, but I am doing a lot of custom tasmota builds.

**Reply by s-hadinger** [reply]
The easy thing we can do is just make an digitalRead on each configured GPIO and report the binary status. For many GPIOs it may not make sense (I2C...) but that's not an issue per se.

**Reply by 0xdefec71f** [reply]
sounds good

**Reply by s-hadinger** [reply]
Working on it

**Reply by Wintespe** [reply]
Please do not change the **_gpio all_** command. In some cases I use it to determine the configuration status. It is better to introduce a new command for purpose above.

**Reply by arendst** [reply]
Try latest dev and command ``GpioRead``

**Reply by 0xdefec71f** [reply]
Works perfectly :-) Thank you!

![image](https://github.com/arendst/Tasmota/assets/35151132/c2388cf3-6ecc-421e-9ddf-dfec57bb8740)
